commit 153123c07dd98d1d86742873b6b4d607fb98e8bf
Author: Peter A. Bigot <pab@peoplepowerco.com>
Date:   Mon Feb 15 21:26:05 2010 -0700

    Revise generic categorization to match binutils and support new architectures.
    
    It appears that the generic MSP430 configurations for gcc (msp1 and msp2)
    are derived from original binutils generic configurations.  Current binutils
    (2.19.1 and 2.20) have six generic configurations, which map to the
    following chip descriptions:
    
    msp430x110:    {"msp1", MSP430_ISA_11, 0}, /* anything without hardware multiply */
    msp430x336:    {"msp2", MSP430_ISA_14, 1}, /* older chips with hardware multiply */
    msp430xG4616:  {"msp3", MSP430_ISA_46, 1},
    msp430x4783:   {"msp4", MSP430_ISA_47, 1},  MSP430 architecture
    msp430x47166:  {"msp5", MSP430_ISA_471, 1}, MSP430X architecture
    msp430x5418:   {"msp6", MSP430_ISA_54, 1},  MSP430XV2 architecture
    
    Next to each in this table is a specific chip that is taken as an example
    for that family.  The first two are the ones used previously in GCC; I don't
    know why msp430x336 was used as the example for msp2 instead of something
    like msp430x147.  The remainder are simply the first listed chip of the
    corresponding category.
    
    Since binutils considers these to be the suite of generic architectures, I
    decided to update the gcc patches to be in sync.  This required changes in
    several places:
    
    msp430.c:msp430_mcu_types was extended with the additional generic types
    
    msp430.h:CPP_SPEC, LINK_SPEC, and CRT_BINUTILS_SPECS were revised to include
      the generics
    
    I also removed the preprocessor definitions that had been added to certain
    lines in CPP_SPEC, and instead placed those within the CPP_mlib_SPEC define
    for the corresponding generic.  This revealed that there are two macros used
    to indicate the presence of a hardware multiply capability.  Where I ended
    up removing chip-specific defines, I verified that the newly assigned
    generic chip matched those defines; I double-checked this against the TI
    headers (e.g., ISA_47 chips are not MSP430X, but ISA_471 are).
    
    Most of this last was aided by a Python script that I used to generate the
    corresponding blocks of code, to make sure specific chips were not dropped
    (e.g., some had been missing from LINK_SPEC).

diff --git a/ports/gcc-4.x/gcc/config/msp430/generate_chip_data.py b/ports/gcc-4.x/gcc/config/msp430/generate_chip_data.py
new file mode 100644
index 0000000..4b25ec7
--- /dev/null
+++ b/ports/gcc-4.x/gcc/config/msp430/generate_chip_data.py
@@ -0,0 +1,241 @@
+text = '''
+	/* F1xx family */
+	{"msp430x110",   MSP430_ISA_11, 0},
+	{"msp430x112",   MSP430_ISA_11, 0},
+
+	{"msp430x1101",  MSP430_ISA_110, 0},
+	{"msp430x1111",  MSP430_ISA_110, 0},
+	{"msp430x1121",  MSP430_ISA_110, 0},
+	{"msp430x1122",  MSP430_ISA_110, 0},
+	{"msp430x1132",  MSP430_ISA_110, 0},
+
+	{"msp430x122",   MSP430_ISA_12, 0},
+	{"msp430x123",   MSP430_ISA_12, 0},
+	{"msp430x1222",  MSP430_ISA_12, 0},
+	{"msp430x1232",  MSP430_ISA_12, 0},
+
+	{"msp430x133",   MSP430_ISA_13, 0},
+	{"msp430x135",   MSP430_ISA_13, 0},
+	{"msp430x1331",  MSP430_ISA_13, 0},
+	{"msp430x1351",  MSP430_ISA_13, 0},
+
+	{"msp430x147",   MSP430_ISA_14, 1},
+	{"msp430x148",   MSP430_ISA_14, 1},
+	{"msp430x149",   MSP430_ISA_14, 1},
+	{"msp430x1471",  MSP430_ISA_14, 1},
+	{"msp430x1481",  MSP430_ISA_14, 1},
+	{"msp430x1491",  MSP430_ISA_14, 1},
+
+	{"msp430x155",   MSP430_ISA_15, 0},
+	{"msp430x156",   MSP430_ISA_15, 0},
+	{"msp430x157",   MSP430_ISA_15, 0},
+
+	{"msp430x167",   MSP430_ISA_16, 1},
+	{"msp430x168",   MSP430_ISA_16, 1},
+	{"msp430x169",   MSP430_ISA_16, 1},
+	{"msp430x1610",  MSP430_ISA_16, 1},
+	{"msp430x1611",  MSP430_ISA_16, 1},
+	{"msp430x1612",  MSP430_ISA_16, 1},
+
+	/* F2xx family */
+	{"msp430x2001",  MSP430_ISA_20, 0},
+	{"msp430x2011",  MSP430_ISA_20, 0},
+
+	{"msp430x2002",  MSP430_ISA_20, 0},
+	{"msp430x2012",  MSP430_ISA_20, 0},
+
+	{"msp430x2003",  MSP430_ISA_20, 0},
+	{"msp430x2013",  MSP430_ISA_20, 0},
+
+	{"msp430x2101",  MSP430_ISA_21, 0},
+	{"msp430x2111",  MSP430_ISA_21, 0},
+	{"msp430x2121",  MSP430_ISA_21, 0},
+	{"msp430x2131",  MSP430_ISA_21, 0},
+
+	{"msp430x2112",  MSP430_ISA_22, 0},
+	{"msp430x2122",  MSP430_ISA_22, 0},
+	{"msp430x2132",  MSP430_ISA_22, 0},
+
+	{"msp430x2232",  MSP430_ISA_22, 0},
+	{"msp430x2252",  MSP430_ISA_22, 0},
+	{"msp430x2272",  MSP430_ISA_22, 0},
+
+	{"msp430x2234",  MSP430_ISA_22, 0},
+	{"msp430x2254",  MSP430_ISA_22, 0},
+	{"msp430x2274",  MSP430_ISA_22, 0},
+
+	{"msp430x233",   MSP430_ISA_23, 1},
+	{"msp430x235",   MSP430_ISA_23, 1},
+
+	{"msp430x2330",  MSP430_ISA_23, 1},
+	{"msp430x2350",  MSP430_ISA_23, 1},
+	{"msp430x2370",  MSP430_ISA_23, 1},
+
+	{"msp430x247",   MSP430_ISA_24, 1},
+	{"msp430x248",   MSP430_ISA_24, 1},
+	{"msp430x249",   MSP430_ISA_24, 1},
+	{"msp430x2410",  MSP430_ISA_24, 1},
+	{"msp430x2471",  MSP430_ISA_24, 1},
+	{"msp430x2481",  MSP430_ISA_24, 1},
+	{"msp430x2491",  MSP430_ISA_24, 1},
+
+	{"msp430x2416",  MSP430_ISA_241, 1},
+	{"msp430x2417",  MSP430_ISA_241, 1},
+	{"msp430x2418",  MSP430_ISA_241, 1},
+	{"msp430x2419",  MSP430_ISA_241, 1},
+
+	{"msp430x2616",  MSP430_ISA_26, 1},
+	{"msp430x2617",  MSP430_ISA_26, 1},
+	{"msp430x2618",  MSP430_ISA_26, 1},
+	{"msp430x2619",  MSP430_ISA_26, 1},
+
+	/* 3xx family (ROM) */
+	{"msp430x311",   MSP430_ISA_31, 0},
+	{"msp430x312",   MSP430_ISA_31, 0},
+	{"msp430x313",   MSP430_ISA_31, 0},
+	{"msp430x314",   MSP430_ISA_31, 0},
+	{"msp430x315",   MSP430_ISA_31, 0},
+
+	{"msp430x323",   MSP430_ISA_32, 0},
+	{"msp430x325",   MSP430_ISA_32, 0},
+
+	{"msp430x336",   MSP430_ISA_33, 1},
+	{"msp430x337",   MSP430_ISA_33, 1},
+
+	/* F4xx family */
+	{"msp430x412",   MSP430_ISA_41, 0},
+	{"msp430x413",   MSP430_ISA_41, 0},
+	{"msp430x415",   MSP430_ISA_41, 0},
+	{"msp430x417",   MSP430_ISA_41, 0},
+
+	{"msp430x423",   MSP430_ISA_42, 1},
+	{"msp430x425",   MSP430_ISA_42, 1},
+	{"msp430x427",   MSP430_ISA_42, 1},
+
+	{"msp430x4250",  MSP430_ISA_42, 0},
+	{"msp430x4260",  MSP430_ISA_42, 0},
+	{"msp430x4270",  MSP430_ISA_42, 0},
+
+	{"msp430xG4250", MSP430_ISA_42, 0},
+	{"msp430xG4260", MSP430_ISA_42, 0},
+	{"msp430xG4270", MSP430_ISA_42, 0},
+
+	{"msp430xE423",  MSP430_ISA_42, 1},
+	{"msp430xE425",  MSP430_ISA_42, 1},
+	{"msp430xE427",  MSP430_ISA_42, 1},
+
+	{"msp430xE4232", MSP430_ISA_42, 1},
+	{"msp430xE4242", MSP430_ISA_42, 1},
+	{"msp430xE4252", MSP430_ISA_42, 1},
+	{"msp430xE4272", MSP430_ISA_42, 1},
+
+	{"msp430xW423",  MSP430_ISA_42, 0},
+	{"msp430xW425",  MSP430_ISA_42, 0},
+	{"msp430xW427",  MSP430_ISA_42, 0},
+
+	{"msp430xG437",  MSP430_ISA_43, 0},
+	{"msp430xG438",  MSP430_ISA_43, 0},
+	{"msp430xG439",  MSP430_ISA_43, 0},
+
+	{"msp430x435",   MSP430_ISA_43, 0},
+	{"msp430x436",   MSP430_ISA_43, 0},
+	{"msp430x437",   MSP430_ISA_43, 0},
+
+	{"msp430x4351",  MSP430_ISA_43, 0},
+	{"msp430x4361",  MSP430_ISA_43, 0},
+	{"msp430x4371",  MSP430_ISA_43, 0},
+
+	{"msp430x447",   MSP430_ISA_44, 1},
+	{"msp430x448",   MSP430_ISA_44, 1},
+	{"msp430x449",   MSP430_ISA_44, 1},
+
+	{"msp430xG4616", MSP430_ISA_46, 1},
+	{"msp430xG4617", MSP430_ISA_46, 1},
+	{"msp430xG4618", MSP430_ISA_46, 1},
+	{"msp430xG4619", MSP430_ISA_46, 1},
+
+	{"msp430x4783",  MSP430_ISA_47, 1},
+	{"msp430x4784",  MSP430_ISA_47, 1},
+	{"msp430x4793",  MSP430_ISA_47, 1},
+	{"msp430x4794",  MSP430_ISA_47, 1},
+        
+	{"msp430x47166", MSP430_ISA_471, 1},
+	{"msp430x47176", MSP430_ISA_471, 1},
+	{"msp430x47186", MSP430_ISA_471, 1},
+	{"msp430x47196", MSP430_ISA_471, 1},
+
+	{"msp430x47167", MSP430_ISA_471, 1},
+	{"msp430x47177", MSP430_ISA_471, 1},
+	{"msp430x47187", MSP430_ISA_471, 1},
+	{"msp430x47197", MSP430_ISA_471, 1},
+
+	/* F5xxx family */
+	{"msp430x5418",  MSP430_ISA_54, 1},
+	{"msp430x5419",  MSP430_ISA_54, 1},
+	{"msp430x5435",  MSP430_ISA_54, 1},
+	{"msp430x5436",  MSP430_ISA_54, 1},
+	{"msp430x5437",  MSP430_ISA_54, 1},
+	{"msp430x5438",  MSP430_ISA_54, 1},
+
+	/* CC430 family */
+	{"cc430x5123",   MSP430_ISA_54, 1},
+	{"cc430x5125",   MSP430_ISA_54, 1},
+	{"cc430x6125",   MSP430_ISA_54, 1},
+	{"cc430x6135",   MSP430_ISA_54, 1},
+	{"cc430x6126",   MSP430_ISA_54, 1},
+	{"cc430x5137",   MSP430_ISA_54, 1},
+	{"cc430x6127",   MSP430_ISA_54, 1},
+	{"cc430x6137",   MSP430_ISA_54, 1},
+'''
+
+import re
+
+pattern_re = re.compile('.*"(.*)".*MSP430_ISA_(\d*),\s*([01])')
+
+exemplary_chip = { 'msp1' : 'msp430x110',
+                   'msp2' : 'msp430x336',
+                   'msp3' : 'msp430xG4616',
+                   'msp4' : 'msp430x4783',
+                   'msp5' : 'msp430x47166',
+                   'msp6' : 'msp430x5418' }
+
+generated =  { }
+seen_chip = { }
+
+for line in text.splitlines():
+    mo = pattern_re.match(line)
+    if mo is None:
+        continue
+    (chip, ctype, hwm) = mo.groups()
+    ctype = int(ctype)
+    hwm = int(hwm)
+    '''
+    if hwm:
+        print 'mmcu?msp2=mmcu?%s' % (chip,)
+    else:
+        print 'mmcu?msp1=mmcu?%s' % (chip,)
+    '''
+    mlib = None
+    if not hwm:
+        mlib = 'msp1'
+    else:
+        if ctype in (26, 241, 46,):
+            mlib = 'msp3'
+        elif ctype in (47,):
+            mlib = 'msp4'
+        elif ctype in (471,):
+            mlib = 'msp5'
+        elif ctype in (54,):
+            mlib = 'msp6'
+        else:
+            mlib = 'msp2'
+    aux_cbs = ''
+    if (exemplary_chip[mlib] == chip):
+        aux_cbs = '|mmcu=%s' % (mlib,)
+    generated.setdefault('MULTILIB_MATCHES', []).append("\tmmcu?%s=mmcu?%s" % (mlib, chip))
+    generated.setdefault('CPP_SPEC', []).append('%%{mmcu=%s:%%(cpp_%s) -D__%s__}' % (chip, mlib, chip.replace('x', '_').upper()))
+    generated.setdefault('LINK_SPEC', []).append('%%{mmcu=%s:-m %s }' % (chip, chip))
+    generated.setdefault('CRT_BINUTILS_SPECS', []).append('%%{mmcu=%s%s:crt%s.o%%s}' % (chip, aux_cbs, chip.replace('msp', '')))
+    
+for (k, v) in generated.items():
+    file(k, 'w').write(" \\\n".join(v))
diff --git a/ports/gcc-4.x/gcc/config/msp430/libgcc.S b/ports/gcc-4.x/gcc/config/msp430/libgcc.S
index 657555b..352ede6 100644
--- a/ports/gcc-4.x/gcc/config/msp430/libgcc.S
+++ b/ports/gcc-4.x/gcc/config/msp430/libgcc.S
@@ -3,6 +3,21 @@
 
 	.section .text.libgcc, "ax", @progbits
 
+#if defined(MSP430_HAS_HWMUL)
+#if defined(__MSP430X2__)
+__MPY=0x4c0
+#else
+__MPY=0x130
+#endif	
+__MPYS=(__MPY   + 0x02)
+__MAC=(__MPY    + 0x04)
+__MACS=(__MPY   + 0x06)
+__OP2=(__MPY    + 0x08)
+__RESLO=(__MPY  + 0x0a)
+__RESHI=(__MPY  + 0x0c)
+__SUMEXT=(__MPY + 0x0e)
+#endif
+
 #if defined (L_cmpdi2)
 	
 	.global	__cmpdi2
@@ -208,15 +223,6 @@ __mulsi3:
 
 #if defined (L_mulsi3hw)
 	
-__MPY=0x130
-__MPYS=0x132
-__MAC=0x134
-__MACS=0x136
-__OP2=0x138
-__RESLO=0x13a
-__RESHI=0x13c
-__SUMEXT=0x13e
-
 	.global	__umulsi3hw
 	.func	__umulsi3hw
 __umulsi3hw:
@@ -1247,15 +1253,6 @@ __muldi3:
 
 #if defined(L_muldi3) && defined(MSP430_HAS_HWMUL)
 
-__MPY=0x130
-__MPYS=0x132
-__MAC=0x134
-__MACS=0x136
-__OP2=0x138
-__RESLO=0x13a
-__RESHI=0x13c
-__SUMEXT=0x13e
-
 #define r_reshh  r11    /* res = arg1 * arg2 */
 #define r_reshl  r10
 #define r_reslh  r9
diff --git a/ports/gcc-4.x/gcc/config/msp430/msp430.c b/ports/gcc-4.x/gcc/config/msp430/msp430.c
index 4626b40..d281e1b 100644
--- a/ports/gcc-4.x/gcc/config/msp430/msp430.c
+++ b/ports/gcc-4.x/gcc/config/msp430/msp430.c
@@ -121,10 +121,17 @@ struct mcu_type_s
 	int has_hwmul;
 };
 
+
+
 static struct mcu_type_s msp430_mcu_types[] = {
-	/* generic types */
-	{"msp1",         MSP430_ISA_1, 0},
-	{"msp2",         MSP430_ISA_2, 1},
+
+	/* Match binutils generic types */
+	{"msp1",         MSP430_ISA_11, 0}, /* anything without hardware multiply */
+	{"msp2",         MSP430_ISA_14, 1}, /* older chips with hardware multiply */
+	{"msp3",         MSP430_ISA_46, 1},
+	{"msp4",         MSP430_ISA_47, 1},
+	{"msp5",         MSP430_ISA_471, 1},
+	{"msp6",         MSP430_ISA_54, 1},
 
 	/* F1xx family */
 	{"msp430x110",   MSP430_ISA_11, 0},
@@ -314,6 +321,7 @@ static struct mcu_type_s msp430_mcu_types[] = {
 	{"cc430x6127",   MSP430_ISA_54, 1},
 	{"cc430x6137",   MSP430_ISA_54, 1},
 
+	/* Add new chips to MULTILIB_MATCHES in t-msp430 as well */
 	{NULL, 0, 0}
 };
 
@@ -526,19 +534,7 @@ rtx gen_rtx_HWREG(const char *name)
 
 void msp430_init_once (void)
 {
-	/******************************
-
-	__MPY=0x130
-	__MPYS=0x132
-	__MAC=0x134
-	__MACS=0x136
-	__OP2=0x138
-	__RESLO=0x13a
-	__RESHI=0x13c
-	__SUMEXT=0x13e
-	__RESSI	<- not natural 
-	*****************************/
-
+	/* Module register addresses chip-specific, added by msp430_file_start */
 	mpy_rtx 	= gen_rtx_HWREG ("__MPY");
 	mpys_rtx 	= gen_rtx_HWREG ("__MPYS");
 	mac_rtx 	= gen_rtx_HWREG ("__MAC");
diff --git a/ports/gcc-4.x/gcc/config/msp430/msp430.h b/ports/gcc-4.x/gcc/config/msp430/msp430.h
index 1c3bd99..da80d83 100644
--- a/ports/gcc-4.x/gcc/config/msp430/msp430.h
+++ b/ports/gcc-4.x/gcc/config/msp430/msp430.h
@@ -2517,7 +2517,11 @@ valid_machine_decl_attribute (DECL, ATTRIBUTES, IDENTIFIER, ARGS)
 
 #define CPP_SPEC "\
 %{!mmcu*|mmcu=msp1:%(cpp_msp1)} \
-%{mmcu=msp2:%(cpp_msp2) -DMSP430_HAS_HWMUL} \
+%{mmcu=msp2:%(cpp_msp2)} \
+%{mmcu=msp3:%(cpp_msp3)} \
+%{mmcu=msp4:%(cpp_msp4)} \
+%{mmcu=msp5:%(cpp_msp5)} \
+%{mmcu=msp6:%(cpp_msp6)} \
 %{mmcu=msp430x110:%(cpp_msp1) -D__MSP430_110__} \
 %{mmcu=msp430x112:%(cpp_msp1) -D__MSP430_112__} \
 %{mmcu=msp430x1101:%(cpp_msp1) -D__MSP430_1101__} \
@@ -2533,21 +2537,21 @@ valid_machine_decl_attribute (DECL, ATTRIBUTES, IDENTIFIER, ARGS)
 %{mmcu=msp430x135:%(cpp_msp1) -D__MSP430_135__} \
 %{mmcu=msp430x1331:%(cpp_msp1) -D__MSP430_1331__} \
 %{mmcu=msp430x1351:%(cpp_msp1) -D__MSP430_1351__} \
-%{mmcu=msp430x147:%(cpp_msp2) -D__MSP430_147__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x148:%(cpp_msp2) -D__MSP430_148__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x149:%(cpp_msp2) -D__MSP430_149__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x1471:%(cpp_msp2) -D__MSP430_1471__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x1481:%(cpp_msp2) -D__MSP430_1481__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x1491:%(cpp_msp2) -D__MSP430_1491__ -DMSP430_HAS_HWMUL} \
+%{mmcu=msp430x147:%(cpp_msp2) -D__MSP430_147__} \
+%{mmcu=msp430x148:%(cpp_msp2) -D__MSP430_148__} \
+%{mmcu=msp430x149:%(cpp_msp2) -D__MSP430_149__} \
+%{mmcu=msp430x1471:%(cpp_msp2) -D__MSP430_1471__} \
+%{mmcu=msp430x1481:%(cpp_msp2) -D__MSP430_1481__} \
+%{mmcu=msp430x1491:%(cpp_msp2) -D__MSP430_1491__} \
 %{mmcu=msp430x155:%(cpp_msp1) -D__MSP430_155__} \
 %{mmcu=msp430x156:%(cpp_msp1) -D__MSP430_156__} \
 %{mmcu=msp430x157:%(cpp_msp1) -D__MSP430_157__} \
-%{mmcu=msp430x167:%(cpp_msp2) -D__MSP430_167__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x168:%(cpp_msp2) -D__MSP430_168__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x169:%(cpp_msp2) -D__MSP430_169__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x1610:%(cpp_msp2) -D__MSP430_1610__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x1611:%(cpp_msp2) -D__MSP430_1611__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x1612:%(cpp_msp2) -D__MSP430_1612__ -DMSP430_HAS_HWMUL} \
+%{mmcu=msp430x167:%(cpp_msp2) -D__MSP430_167__} \
+%{mmcu=msp430x168:%(cpp_msp2) -D__MSP430_168__} \
+%{mmcu=msp430x169:%(cpp_msp2) -D__MSP430_169__} \
+%{mmcu=msp430x1610:%(cpp_msp2) -D__MSP430_1610__} \
+%{mmcu=msp430x1611:%(cpp_msp2) -D__MSP430_1611__} \
+%{mmcu=msp430x1612:%(cpp_msp2) -D__MSP430_1612__} \
 %{mmcu=msp430x2001:%(cpp_msp1) -D__MSP430_2001__} \
 %{mmcu=msp430x2011:%(cpp_msp1) -D__MSP430_2011__} \
 %{mmcu=msp430x2002:%(cpp_msp1) -D__MSP430_2002__} \
@@ -2567,26 +2571,26 @@ valid_machine_decl_attribute (DECL, ATTRIBUTES, IDENTIFIER, ARGS)
 %{mmcu=msp430x2234:%(cpp_msp1) -D__MSP430_2234__} \
 %{mmcu=msp430x2254:%(cpp_msp1) -D__MSP430_2254__} \
 %{mmcu=msp430x2274:%(cpp_msp1) -D__MSP430_2274__} \
-%{mmcu=msp430x233:%(cpp_msp2) -D__MSP430_233__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x235:%(cpp_msp2) -D__MSP430_235__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2330:%(cpp_msp2) -D__MSP430_2330__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2350:%(cpp_msp2) -D__MSP430_2350__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2370:%(cpp_msp2) -D__MSP430_2370__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x247:%(cpp_msp2) -D__MSP430_247__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x248:%(cpp_msp2) -D__MSP430_248__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x249:%(cpp_msp2) -D__MSP430_249__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2410:%(cpp_msp2) -D__MSP430_2410__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2471:%(cpp_msp2) -D__MSP430_2471__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2481:%(cpp_msp2) -D__MSP430_2481__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2491:%(cpp_msp2) -D__MSP430_2491__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2416:%(cpp_msp2) -D__MSP430_2416__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2417:%(cpp_msp2) -D__MSP430_2417__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2418:%(cpp_msp2) -D__MSP430_2418__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2419:%(cpp_msp2) -D__MSP430_2419__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2616:%(cpp_msp2) -D__MSP430_2616__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2617:%(cpp_msp2) -D__MSP430_2617__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2618:%(cpp_msp2) -D__MSP430_2618__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x2619:%(cpp_msp2) -D__MSP430_2619__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
+%{mmcu=msp430x233:%(cpp_msp2) -D__MSP430_233__} \
+%{mmcu=msp430x235:%(cpp_msp2) -D__MSP430_235__} \
+%{mmcu=msp430x2330:%(cpp_msp2) -D__MSP430_2330__} \
+%{mmcu=msp430x2350:%(cpp_msp2) -D__MSP430_2350__} \
+%{mmcu=msp430x2370:%(cpp_msp2) -D__MSP430_2370__} \
+%{mmcu=msp430x247:%(cpp_msp2) -D__MSP430_247__} \
+%{mmcu=msp430x248:%(cpp_msp2) -D__MSP430_248__} \
+%{mmcu=msp430x249:%(cpp_msp2) -D__MSP430_249__} \
+%{mmcu=msp430x2410:%(cpp_msp2) -D__MSP430_2410__} \
+%{mmcu=msp430x2471:%(cpp_msp2) -D__MSP430_2471__} \
+%{mmcu=msp430x2481:%(cpp_msp2) -D__MSP430_2481__} \
+%{mmcu=msp430x2491:%(cpp_msp2) -D__MSP430_2491__} \
+%{mmcu=msp430x2416:%(cpp_msp3) -D__MSP430_2416__} \
+%{mmcu=msp430x2417:%(cpp_msp3) -D__MSP430_2417__} \
+%{mmcu=msp430x2418:%(cpp_msp3) -D__MSP430_2418__} \
+%{mmcu=msp430x2419:%(cpp_msp3) -D__MSP430_2419__} \
+%{mmcu=msp430x2616:%(cpp_msp3) -D__MSP430_2616__} \
+%{mmcu=msp430x2617:%(cpp_msp3) -D__MSP430_2617__} \
+%{mmcu=msp430x2618:%(cpp_msp3) -D__MSP430_2618__} \
+%{mmcu=msp430x2619:%(cpp_msp3) -D__MSP430_2619__} \
 %{mmcu=msp430x311:%(cpp_msp1) -D__MSP430_311__} \
 %{mmcu=msp430x312:%(cpp_msp1) -D__MSP430_312__} \
 %{mmcu=msp430x313:%(cpp_msp1) -D__MSP430_313__} \
@@ -2594,28 +2598,28 @@ valid_machine_decl_attribute (DECL, ATTRIBUTES, IDENTIFIER, ARGS)
 %{mmcu=msp430x315:%(cpp_msp1) -D__MSP430_315__} \
 %{mmcu=msp430x323:%(cpp_msp1) -D__MSP430_323__} \
 %{mmcu=msp430x325:%(cpp_msp1) -D__MSP430_325__} \
-%{mmcu=msp430x336:%(cpp_msp2) -D__MSP430_336__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x337:%(cpp_msp2) -D__MSP430_337__ -DMSP430_HAS_HWMUL} \
+%{mmcu=msp430x336:%(cpp_msp2) -D__MSP430_336__} \
+%{mmcu=msp430x337:%(cpp_msp2) -D__MSP430_337__} \
 %{mmcu=msp430x412:%(cpp_msp1) -D__MSP430_412__} \
 %{mmcu=msp430x413:%(cpp_msp1) -D__MSP430_413__} \
 %{mmcu=msp430x415:%(cpp_msp1) -D__MSP430_415__} \
 %{mmcu=msp430x417:%(cpp_msp1) -D__MSP430_417__} \
-%{mmcu=msp430x423:%(cpp_msp2) -D__MSP430_423__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x425:%(cpp_msp2) -D__MSP430_425__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x427:%(cpp_msp2) -D__MSP430_427__ -DMSP430_HAS_HWMUL} \
+%{mmcu=msp430x423:%(cpp_msp2) -D__MSP430_423__} \
+%{mmcu=msp430x425:%(cpp_msp2) -D__MSP430_425__} \
+%{mmcu=msp430x427:%(cpp_msp2) -D__MSP430_427__} \
 %{mmcu=msp430x4250:%(cpp_msp1) -D__MSP430_4250__} \
 %{mmcu=msp430x4260:%(cpp_msp1) -D__MSP430_4260__} \
 %{mmcu=msp430x4270:%(cpp_msp1) -D__MSP430_4270__} \
 %{mmcu=msp430xG4250:%(cpp_msp1) -D__MSP430_G4250__} \
 %{mmcu=msp430xG4260:%(cpp_msp1) -D__MSP430_G4260__} \
 %{mmcu=msp430xG4270:%(cpp_msp1) -D__MSP430_G4270__} \
-%{mmcu=msp430xE423:%(cpp_msp2) -D__MSP430_E423__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430xE425:%(cpp_msp2) -D__MSP430_E425__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430xE427:%(cpp_msp2) -D__MSP430_E427__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430xE4232:%(cpp_msp2) -D__MSP430_E4232__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430xE4242:%(cpp_msp2) -D__MSP430_E4242__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430xE4252:%(cpp_msp2) -D__MSP430_E4252__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430xE4272:%(cpp_msp2) -D__MSP430_E4272__ -DMSP430_HAS_HWMUL} \
+%{mmcu=msp430xE423:%(cpp_msp2) -D__MSP430_E423__} \
+%{mmcu=msp430xE425:%(cpp_msp2) -D__MSP430_E425__} \
+%{mmcu=msp430xE427:%(cpp_msp2) -D__MSP430_E427__} \
+%{mmcu=msp430xE4232:%(cpp_msp2) -D__MSP430_E4232__} \
+%{mmcu=msp430xE4242:%(cpp_msp2) -D__MSP430_E4242__} \
+%{mmcu=msp430xE4252:%(cpp_msp2) -D__MSP430_E4252__} \
+%{mmcu=msp430xE4272:%(cpp_msp2) -D__MSP430_E4272__} \
 %{mmcu=msp430xW423:%(cpp_msp1) -D__MSP430_W423__} \
 %{mmcu=msp430xW425:%(cpp_msp1) -D__MSP430_W425__} \
 %{mmcu=msp430xW427:%(cpp_msp1) -D__MSP430_W427__} \
@@ -2628,39 +2632,39 @@ valid_machine_decl_attribute (DECL, ATTRIBUTES, IDENTIFIER, ARGS)
 %{mmcu=msp430x4351:%(cpp_msp1) -D__MSP430_4351__} \
 %{mmcu=msp430x4361:%(cpp_msp1) -D__MSP430_4361__} \
 %{mmcu=msp430x4371:%(cpp_msp1) -D__MSP430_4371__} \
-%{mmcu=msp430x447:%(cpp_msp2) -D__MSP430_447__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x448:%(cpp_msp2) -D__MSP430_448__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x449:%(cpp_msp2) -D__MSP430_449__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430xG4616:%(cpp_msp2) -D__MSP430_G4616__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430xG4617:%(cpp_msp2) -D__MSP430_G4617__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430xG4618:%(cpp_msp2) -D__MSP430_G4618__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430xG4619:%(cpp_msp2) -D__MSP430_G4619__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x4783:%(cpp_msp2) -D__MSP430_4783__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x4784:%(cpp_msp2) -D__MSP430_4784__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x4793:%(cpp_msp2) -D__MSP430_4793__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x4794:%(cpp_msp2) -D__MSP430_4794__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x47166:%(cpp_msp2) -D__MSP430_47166__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x47176:%(cpp_msp2) -D__MSP430_47176__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x47186:%(cpp_msp2) -D__MSP430_47186__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x47196:%(cpp_msp2) -D__MSP430_47196__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x47167:%(cpp_msp2) -D__MSP430_47167__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x47177:%(cpp_msp2) -D__MSP430_47177__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x47187:%(cpp_msp2) -D__MSP430_47187__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x47197:%(cpp_msp2) -D__MSP430_47197__ -D__MSP430X__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x5418:%(cpp_msp2) -D__MSP430_5418__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x5419:%(cpp_msp2) -D__MSP430_5419__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x5435:%(cpp_msp2) -D__MSP430_5435__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x5436:%(cpp_msp2) -D__MSP430_5436__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x5437:%(cpp_msp2) -D__MSP430_5437__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=msp430x5438:%(cpp_msp2) -D__MSP430_5438__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=cc430x5123:%(cpp_msp2) -D__CC430_5123__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=cc430x5125:%(cpp_msp2) -D__CC430_5125__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=cc430x6125:%(cpp_msp2) -D__CC430_6125__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=cc430x6135:%(cpp_msp2) -D__CC430_6135__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=cc430x6126:%(cpp_msp2) -D__CC430_6126__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=cc430x5137:%(cpp_msp2) -D__CC430_5137__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=cc430x6127:%(cpp_msp2) -D__CC430_6127__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
-%{mmcu=cc430x6137:%(cpp_msp2) -D__CC430_6137__ -D__MSP430X2__ -DMSP430_HAS_HWMUL} \
+%{mmcu=msp430x447:%(cpp_msp2) -D__MSP430_447__} \
+%{mmcu=msp430x448:%(cpp_msp2) -D__MSP430_448__} \
+%{mmcu=msp430x449:%(cpp_msp2) -D__MSP430_449__} \
+%{mmcu=msp430xG4616:%(cpp_msp3) -D__MSP430_G4616__} \
+%{mmcu=msp430xG4617:%(cpp_msp3) -D__MSP430_G4617__} \
+%{mmcu=msp430xG4618:%(cpp_msp3) -D__MSP430_G4618__} \
+%{mmcu=msp430xG4619:%(cpp_msp3) -D__MSP430_G4619__} \
+%{mmcu=msp430x4783:%(cpp_msp4) -D__MSP430_4783__} \
+%{mmcu=msp430x4784:%(cpp_msp4) -D__MSP430_4784__} \
+%{mmcu=msp430x4793:%(cpp_msp4) -D__MSP430_4793__} \
+%{mmcu=msp430x4794:%(cpp_msp4) -D__MSP430_4794__} \
+%{mmcu=msp430x47166:%(cpp_msp5) -D__MSP430_47166__} \
+%{mmcu=msp430x47176:%(cpp_msp5) -D__MSP430_47176__} \
+%{mmcu=msp430x47186:%(cpp_msp5) -D__MSP430_47186__} \
+%{mmcu=msp430x47196:%(cpp_msp5) -D__MSP430_47196__} \
+%{mmcu=msp430x47167:%(cpp_msp5) -D__MSP430_47167__} \
+%{mmcu=msp430x47177:%(cpp_msp5) -D__MSP430_47177__} \
+%{mmcu=msp430x47187:%(cpp_msp5) -D__MSP430_47187__} \
+%{mmcu=msp430x47197:%(cpp_msp5) -D__MSP430_47197__} \
+%{mmcu=msp430x5418:%(cpp_msp6) -D__MSP430_5418__} \
+%{mmcu=msp430x5419:%(cpp_msp6) -D__MSP430_5419__} \
+%{mmcu=msp430x5435:%(cpp_msp6) -D__MSP430_5435__} \
+%{mmcu=msp430x5436:%(cpp_msp6) -D__MSP430_5436__} \
+%{mmcu=msp430x5437:%(cpp_msp6) -D__MSP430_5437__} \
+%{mmcu=msp430x5438:%(cpp_msp6) -D__MSP430_5438__} \
+%{mmcu=cc430x5123:%(cpp_msp6) -D__CC430_5123__} \
+%{mmcu=cc430x5125:%(cpp_msp6) -D__CC430_5125__} \
+%{mmcu=cc430x6125:%(cpp_msp6) -D__CC430_6125__} \
+%{mmcu=cc430x6135:%(cpp_msp6) -D__CC430_6135__} \
+%{mmcu=cc430x6126:%(cpp_msp6) -D__CC430_6126__} \
+%{mmcu=cc430x5137:%(cpp_msp6) -D__CC430_5137__} \
+%{mmcu=cc430x6127:%(cpp_msp6) -D__CC430_6127__} \
+%{mmcu=cc430x6137:%(cpp_msp6) -D__CC430_6137__} \
 %{posix:-D_POSIX_SOURCE} %{mIAR:-D_IAR_ASSEMBLER_}"
 
 /* A C string constant that tells the GNU CC driver program options to
@@ -2698,7 +2702,11 @@ valid_machine_decl_attribute (DECL, ATTRIBUTES, IDENTIFIER, ARGS)
 #define LINK_SPEC "\
 %{!mmcu*:-m msp430x110} \
 %{mmcu=msp1:-m msp430x110} \
-%{mmcu=msp2:-m msp430x336} \
+%{mmcu=msp2:-m msp430x147} \
+%{mmcu=msp3:-m msp430xG4616} \
+%{mmcu=msp4:-m msp430x4783} \
+%{mmcu=msp5:-m msp430x47166} \
+%{mmcu=msp6:-m msp430x5418} \
 %{mmcu=msp430x110:-m msp430x110 } \
 %{mmcu=msp430x112:-m msp430x112 } \
 %{mmcu=msp430x1101:-m msp430x1101 } \
@@ -2806,6 +2814,9 @@ valid_machine_decl_attribute (DECL, ATTRIBUTES, IDENTIFIER, ARGS)
 %{mmcu=msp430x435:-m msp430x435 } \
 %{mmcu=msp430x436:-m msp430x436 } \
 %{mmcu=msp430x437:-m msp430x437 } \
+%{mmcu=msp430x4351:-m msp430x4351 } \
+%{mmcu=msp430x4361:-m msp430x4361 } \
+%{mmcu=msp430x4371:-m msp430x4371 } \
 %{mmcu=msp430x447:-m msp430x447 } \
 %{mmcu=msp430x448:-m msp430x448 } \
 %{mmcu=msp430x449:-m msp430x449 } \
@@ -2831,14 +2842,14 @@ valid_machine_decl_attribute (DECL, ATTRIBUTES, IDENTIFIER, ARGS)
 %{mmcu=msp430x5436:-m msp430x5436 } \
 %{mmcu=msp430x5437:-m msp430x5437 } \
 %{mmcu=msp430x5438:-m msp430x5438 } \
-%{mmcu=cc430x5123: -m cc430x5123 } \
-%{mmcu=cc430x5125: -m cc430x5125 } \
-%{mmcu=cc430x6125: -m cc430x6125 } \
-%{mmcu=cc430x6135: -m cc430x6135 } \
-%{mmcu=cc430x6126: -m cc430x6126 } \
-%{mmcu=cc430x5137: -m cc430x5137 } \
-%{mmcu=cc430x6127: -m cc430x6127 } \
-%{mmcu=cc430x6137: -m cc430x6137 } \
+%{mmcu=cc430x5123:-m cc430x5123 } \
+%{mmcu=cc430x5125:-m cc430x5125 } \
+%{mmcu=cc430x6125:-m cc430x6125 } \
+%{mmcu=cc430x6135:-m cc430x6135 } \
+%{mmcu=cc430x6126:-m cc430x6126 } \
+%{mmcu=cc430x5137:-m cc430x5137 } \
+%{mmcu=cc430x6127:-m cc430x6127 } \
+%{mmcu=cc430x6137:-m cc430x6137 } \
 "
 
 #define LIB_SPEC "-lc"
@@ -2974,16 +2985,24 @@ valid_machine_decl_attribute (DECL, ATTRIBUTES, IDENTIFIER, ARGS)
 %{mmcu=msp430x447:crt430x447.o%s} \
 %{mmcu=msp430x448:crt430x448.o%s} \
 %{mmcu=msp430x449:crt430x449.o%s} \
-%{mmcu=msp430xG4616:crt430xG4616.o%s} \
+%{mmcu=msp430xG4616|mmcu=msp3:crt430xG4616.o%s} \
 %{mmcu=msp430xG4617:crt430xG4617.o%s} \
 %{mmcu=msp430xG4618:crt430xG4618.o%s} \
-%{mmcu=msp430xG4619:crt430xG4619.o%s}\
-%{mmcu=msp430x4783:crt430x4783.o%s} \
+%{mmcu=msp430xG4619:crt430xG4619.o%s} \
+%{mmcu=msp430x4783|mmcu=msp4:crt430x4783.o%s} \
 %{mmcu=msp430x4784:crt430x4784.o%s} \
 %{mmcu=msp430x4793:crt430x4793.o%s} \
 %{mmcu=msp430x4794:crt430x4794.o%s} \
-%{mmcu=msp430x5418:crt430x5418.o%s} \
-%{mmcu=msp430x5419:crt430x5419.o%s}\
+%{mmcu=msp430x47166|mmcu=msp5:crt430x47166.o%s} \
+%{mmcu=msp430x47176:crt430x47176.o%s} \
+%{mmcu=msp430x47186:crt430x47186.o%s} \
+%{mmcu=msp430x47196:crt430x47196.o%s} \
+%{mmcu=msp430x47167:crt430x47167.o%s} \
+%{mmcu=msp430x47177:crt430x47177.o%s} \
+%{mmcu=msp430x47187:crt430x47187.o%s} \
+%{mmcu=msp430x47197:crt430x47197.o%s} \
+%{mmcu=msp430x5418|mmcu=msp6:crt430x5418.o%s} \
+%{mmcu=msp430x5419:crt430x5419.o%s} \
 %{mmcu=msp430x5435:crt430x5435.o%s} \
 %{mmcu=msp430x5436:crt430x5436.o%s} \
 %{mmcu=msp430x5437:crt430x5437.o%s} \
@@ -2998,14 +3017,20 @@ valid_machine_decl_attribute (DECL, ATTRIBUTES, IDENTIFIER, ARGS)
 %{mmcu=cc430x6137:crtcc430x6137.o%s} \
 "
 
-
-
 #define CPP_MSP1_SPEC " -DMSP430_NO_HW_MUL "
-#define CPP_MSP2_SPEC " -DMSP430_HAS_HW_MUL "
+#define CPP_MSP2_SPEC " -DMSP430_HAS_HW_MUL -DMSP430_HAS_HWMUL "
+#define CPP_MSP3_SPEC " -DMSP430_HAS_HW_MUL -DMSP430_HAS_HWMUL -D__MSP430X__"
+#define CPP_MSP4_SPEC " -DMSP430_HAS_HW_MUL -DMSP430_HAS_HWMUL"
+#define CPP_MSP5_SPEC " -DMSP430_HAS_HW_MUL -DMSP430_HAS_HWMUL -D__MSP430X__"
+#define CPP_MSP6_SPEC " -DMSP430_HAS_HW_MUL -DMSP430_HAS_HWMUL -D__MSP430X2__"
 
 #define EXTRA_SPECS                           	\
 {"cpp_msp1",CPP_MSP1_SPEC},			\
 {"cpp_msp2",CPP_MSP2_SPEC},			\
+{"cpp_msp3",CPP_MSP3_SPEC},			\
+{"cpp_msp4",CPP_MSP4_SPEC},			\
+{"cpp_msp5",CPP_MSP5_SPEC},			\
+{"cpp_msp6",CPP_MSP6_SPEC},			\
 {"crt_binutils", CRT_BINUTILS_SPECS},
 
 /* Define this macro to provide additional specifications to put in
diff --git a/ports/gcc-4.x/gcc/config/msp430/t-msp430 b/ports/gcc-4.x/gcc/config/msp430/t-msp430
index 5365ef3..6480877 100644
--- a/ports/gcc-4.x/gcc/config/msp430/t-msp430
+++ b/ports/gcc-4.x/gcc/config/msp430/t-msp430
@@ -61,70 +61,153 @@ fp-bit.c: $(srcdir)/config/fp-bit.c $(srcdir)/config/msp430/t-msp430
 
 FPBIT = fp-bit.c
 
-MULTILIB_OPTIONS = mmcu=msp1/mmcu=msp2
-MULTILIB_DIRNAMES = msp1 msp2
-
+MULTILIB_OPTIONS = mmcu=msp1/mmcu=msp2/mmcu=msp3/mmcu=msp4/mmcu=msp5/mmcu=msp6
+MULTILIB_DIRNAMES = msp1 msp2 msp3 msp4 msp5 msp6
 
 MULTILIB_MATCHES = \
-	mmcu?msp1=mmcu?msp430x110   mmcu?msp1=mmcu?msp430x112 \
-	mmcu?msp1=mmcu?msp430x1101  mmcu?msp1=mmcu?msp430x1111  mmcu?msp1=mmcu?msp430x1121 \
-	mmcu?msp1=mmcu?msp430x1122  mmcu?msp1=mmcu?msp430x1132 \
-	mmcu?msp1=mmcu?msp430x122   mmcu?msp1=mmcu?msp430x123 \
-	mmcu?msp1=mmcu?msp430x1222  mmcu?msp1=mmcu?msp430x1232 \
-	mmcu?msp1=mmcu?msp430x133   mmcu?msp1=mmcu?msp430x135 \
-	mmcu?msp1=mmcu?msp430x1331  mmcu?msp1=mmcu?msp430x1351 \
-	mmcu?msp2=mmcu?msp430x147   mmcu?msp2=mmcu?msp430x148   mmcu?msp2=mmcu?msp430x149 \
-	mmcu?msp2=mmcu?msp430x1471  mmcu?msp2=mmcu?msp430x1481  mmcu?msp2=mmcu?msp430x1491 \
-	mmcu?msp1=mmcu?msp430x155   mmcu?msp1=mmcu?msp430x156   mmcu?msp1=mmcu?msp430x157 \
-	mmcu?msp2=mmcu?msp430x167   mmcu?msp2=mmcu?msp430x168   mmcu?msp2=mmcu?msp430x169 \
-	mmcu?msp2=mmcu?msp430x1610  mmcu?msp2=mmcu?msp430x1611  mmcu?msp2=mmcu?msp430x1612 \
-	mmcu?msp1=mmcu?msp430x2001  mmcu?msp1=mmcu?msp430x2011 \
-	mmcu?msp1=mmcu?msp430x2002  mmcu?msp1=mmcu?msp430x2012 \
-	mmcu?msp1=mmcu?msp430x2003  mmcu?msp1=mmcu?msp430x2013 \
-	mmcu?msp1=mmcu?msp430x2101  mmcu?msp1=mmcu?msp430x2111  mmcu?msp1=mmcu?msp430x2121 \
+	mmcu?msp1=mmcu?msp430x110 \
+	mmcu?msp1=mmcu?msp430x112 \
+	mmcu?msp1=mmcu?msp430x1101 \
+	mmcu?msp1=mmcu?msp430x1111 \
+	mmcu?msp1=mmcu?msp430x1121 \
+	mmcu?msp1=mmcu?msp430x1122 \
+	mmcu?msp1=mmcu?msp430x1132 \
+	mmcu?msp1=mmcu?msp430x122 \
+	mmcu?msp1=mmcu?msp430x123 \
+	mmcu?msp1=mmcu?msp430x1222 \
+	mmcu?msp1=mmcu?msp430x1232 \
+	mmcu?msp1=mmcu?msp430x133 \
+	mmcu?msp1=mmcu?msp430x135 \
+	mmcu?msp1=mmcu?msp430x1331 \
+	mmcu?msp1=mmcu?msp430x1351 \
+	mmcu?msp2=mmcu?msp430x147 \
+	mmcu?msp2=mmcu?msp430x148 \
+	mmcu?msp2=mmcu?msp430x149 \
+	mmcu?msp2=mmcu?msp430x1471 \
+	mmcu?msp2=mmcu?msp430x1481 \
+	mmcu?msp2=mmcu?msp430x1491 \
+	mmcu?msp1=mmcu?msp430x155 \
+	mmcu?msp1=mmcu?msp430x156 \
+	mmcu?msp1=mmcu?msp430x157 \
+	mmcu?msp2=mmcu?msp430x167 \
+	mmcu?msp2=mmcu?msp430x168 \
+	mmcu?msp2=mmcu?msp430x169 \
+	mmcu?msp2=mmcu?msp430x1610 \
+	mmcu?msp2=mmcu?msp430x1611 \
+	mmcu?msp2=mmcu?msp430x1612 \
+	mmcu?msp1=mmcu?msp430x2001 \
+	mmcu?msp1=mmcu?msp430x2011 \
+	mmcu?msp1=mmcu?msp430x2002 \
+	mmcu?msp1=mmcu?msp430x2012 \
+	mmcu?msp1=mmcu?msp430x2003 \
+	mmcu?msp1=mmcu?msp430x2013 \
+	mmcu?msp1=mmcu?msp430x2101 \
+	mmcu?msp1=mmcu?msp430x2111 \
+	mmcu?msp1=mmcu?msp430x2121 \
 	mmcu?msp1=mmcu?msp430x2131 \
-	mmcu?msp1=mmcu?msp430x2112  mmcu?msp1=mmcu?msp430x2122 \
-	mmcu?msp1=mmcu?msp430x2232  mmcu?msp1=mmcu?msp430x2252  mmcu?msp1=mmcu?msp430x2272 \
-	mmcu?msp1=mmcu?msp430x2234  mmcu?msp1=mmcu?msp430x2254  mmcu?msp1=mmcu?msp430x2274 \
-	mmcu?msp2=mmcu?msp430x233   mmcu?msp2=mmcu?msp430x235 \
-	mmcu?msp2=mmcu?msp430x2330  mmcu?msp2=mmcu?msp430x2350  mmcu?msp2=mmcu?msp430x2370 \
-	mmcu?msp2=mmcu?msp430x247   mmcu?msp2=mmcu?msp430x248   mmcu?msp2=mmcu?msp430x249 \
+	mmcu?msp1=mmcu?msp430x2112 \
+	mmcu?msp1=mmcu?msp430x2122 \
+	mmcu?msp1=mmcu?msp430x2132 \
+	mmcu?msp1=mmcu?msp430x2232 \
+	mmcu?msp1=mmcu?msp430x2252 \
+	mmcu?msp1=mmcu?msp430x2272 \
+	mmcu?msp1=mmcu?msp430x2234 \
+	mmcu?msp1=mmcu?msp430x2254 \
+	mmcu?msp1=mmcu?msp430x2274 \
+	mmcu?msp2=mmcu?msp430x233 \
+	mmcu?msp2=mmcu?msp430x235 \
+	mmcu?msp2=mmcu?msp430x2330 \
+	mmcu?msp2=mmcu?msp430x2350 \
+	mmcu?msp2=mmcu?msp430x2370 \
+	mmcu?msp2=mmcu?msp430x247 \
+	mmcu?msp2=mmcu?msp430x248 \
+	mmcu?msp2=mmcu?msp430x249 \
 	mmcu?msp2=mmcu?msp430x2410 \
-	mmcu?msp2=mmcu?msp430x2471  mmcu?msp2=mmcu?msp430x2481  mmcu?msp2=mmcu?msp430x2491 \
-	mmcu?msp2=mmcu?msp430x2416  mmcu?msp2=mmcu?msp430x2417  mmcu?msp2=mmcu?msp430x2418 \
-	mmcu?msp2=mmcu?msp430x2419 \
-	mmcu?msp2=mmcu?msp430x2616  mmcu?msp2=mmcu?msp430x2617  mmcu?msp2=mmcu?msp430x2618 \
-	mmcu?msp2=mmcu?msp430x2619 \
-	mmcu?msp1=mmcu?msp430x311   mmcu?msp1=mmcu?msp430x312   mmcu?msp1=mmcu?msp430x313 \
-	mmcu?msp1=mmcu?msp430x314   mmcu?msp1=mmcu?msp430x315 \
-	mmcu?msp1=mmcu?msp430x323   mmcu?msp1=mmcu?msp430x325 \
-	mmcu?msp2=mmcu?msp430x336   mmcu?msp2=mmcu?msp430x337 \
-	mmcu?msp1=mmcu?msp430x412   mmcu?msp1=mmcu?msp430x413 \
-	mmcu?msp1=mmcu?msp430x415   mmcu?msp1=mmcu?msp430x417 \
-	mmcu?msp2=mmcu?msp430x423   mmcu?msp2=mmcu?msp430x425   mmcu?msp2=mmcu?msp430x427 \
-	mmcu?msp1=mmcu?msp430x4250  mmcu?msp1=mmcu?msp430x4260  mmcu?msp1=mmcu?msp430x4270 \
-	mmcu?msp1=mmcu?msp430xG4250 mmcu?msp1=mmcu?msp430xG4260 mmcu?msp1=mmcu?msp430xG4270 \
-	mmcu?msp2=mmcu?msp430xE423  mmcu?msp2=mmcu?msp430xE425  mmcu?msp2=mmcu?msp430xE427 \
-	mmcu?msp2=mmcu?msp430xE4232 mmcu?msp2=mmcu?msp430xE4242 mmcu?msp2=mmcu?msp430xE4252 \
+	mmcu?msp2=mmcu?msp430x2471 \
+	mmcu?msp2=mmcu?msp430x2481 \
+	mmcu?msp2=mmcu?msp430x2491 \
+	mmcu?msp3=mmcu?msp430x2416 \
+	mmcu?msp3=mmcu?msp430x2417 \
+	mmcu?msp3=mmcu?msp430x2418 \
+	mmcu?msp3=mmcu?msp430x2419 \
+	mmcu?msp3=mmcu?msp430x2616 \
+	mmcu?msp3=mmcu?msp430x2617 \
+	mmcu?msp3=mmcu?msp430x2618 \
+	mmcu?msp3=mmcu?msp430x2619 \
+	mmcu?msp1=mmcu?msp430x311 \
+	mmcu?msp1=mmcu?msp430x312 \
+	mmcu?msp1=mmcu?msp430x313 \
+	mmcu?msp1=mmcu?msp430x314 \
+	mmcu?msp1=mmcu?msp430x315 \
+	mmcu?msp1=mmcu?msp430x323 \
+	mmcu?msp1=mmcu?msp430x325 \
+	mmcu?msp2=mmcu?msp430x336 \
+	mmcu?msp2=mmcu?msp430x337 \
+	mmcu?msp1=mmcu?msp430x412 \
+	mmcu?msp1=mmcu?msp430x413 \
+	mmcu?msp1=mmcu?msp430x415 \
+	mmcu?msp1=mmcu?msp430x417 \
+	mmcu?msp2=mmcu?msp430x423 \
+	mmcu?msp2=mmcu?msp430x425 \
+	mmcu?msp2=mmcu?msp430x427 \
+	mmcu?msp1=mmcu?msp430x4250 \
+	mmcu?msp1=mmcu?msp430x4260 \
+	mmcu?msp1=mmcu?msp430x4270 \
+	mmcu?msp1=mmcu?msp430xG4250 \
+	mmcu?msp1=mmcu?msp430xG4260 \
+	mmcu?msp1=mmcu?msp430xG4270 \
+	mmcu?msp2=mmcu?msp430xE423 \
+	mmcu?msp2=mmcu?msp430xE425 \
+	mmcu?msp2=mmcu?msp430xE427 \
+	mmcu?msp2=mmcu?msp430xE4232 \
+	mmcu?msp2=mmcu?msp430xE4242 \
+	mmcu?msp2=mmcu?msp430xE4252 \
 	mmcu?msp2=mmcu?msp430xE4272 \
-	mmcu?msp1=mmcu?msp430xW423  mmcu?msp1=mmcu?msp430xW425  mmcu?msp1=mmcu?msp430xW427 \
-	mmcu?msp1=mmcu?msp430xG437  mmcu?msp1=mmcu?msp430xG438  mmcu?msp1=mmcu?msp430xG439 \
-	mmcu?msp1=mmcu?msp430x435   mmcu?msp1=mmcu?msp430x436   mmcu?msp1=mmcu?msp430x437 \
-	mmcu?msp1=mmcu?msp430x4351  mmcu?msp1=mmcu?msp430x4361  mmcu?msp1=mmcu?msp430x4371 \
-	mmcu?msp2=mmcu?msp430x447   mmcu?msp2=mmcu?msp430x448   mmcu?msp2=mmcu?msp430x449 \
-	mmcu?msp2=mmcu?msp430xG4616 mmcu?msp2=mmcu?msp430xG4617 mmcu?msp2=mmcu?msp430xG4618 \
-	mmcu?msp2=mmcu?msp430xG4619 \
-	mmcu?msp2=mmcu?msp430x4783  mmcu?msp2=mmcu?msp430x4784  mmcu?msp2=mmcu?msp430x4793 \
-	mmcu?msp2=mmcu?msp430x4794 \
-	mmcu?msp2=mmcu?msp430x5418 mmcu?msp2=mmcu?msp430x5419 \
-	mmcu?msp2=mmcu?msp430x5435 mmcu?msp2=mmcu?msp430x5436 mmcu?msp2=mmcu?msp430x5437 \
-	mmcu?msp2=mmcu?msp430x5438 \
-	mmcu?msp2=mmcu?cc430x5123 \
-	mmcu?msp2=mmcu?cc430x5125 mmcu?msp2=mmcu?cc430x6125 mmcu?msp2=mmcu?cc430x6135 \
-	mmcu?msp2=mmcu?cc430x6126 \
-	mmcu?msp2=mmcu?cc430x5137 mmcu?msp2=mmcu?cc430x6127 mmcu?msp2=mmcu?cc430x6137
-#todo: 47xx has 32*32 HW MUL
-#todo: 54xx has 32*32 HW MUL
+	mmcu?msp1=mmcu?msp430xW423 \
+	mmcu?msp1=mmcu?msp430xW425 \
+	mmcu?msp1=mmcu?msp430xW427 \
+	mmcu?msp1=mmcu?msp430xG437 \
+	mmcu?msp1=mmcu?msp430xG438 \
+	mmcu?msp1=mmcu?msp430xG439 \
+	mmcu?msp1=mmcu?msp430x435 \
+	mmcu?msp1=mmcu?msp430x436 \
+	mmcu?msp1=mmcu?msp430x437 \
+	mmcu?msp1=mmcu?msp430x4351 \
+	mmcu?msp1=mmcu?msp430x4361 \
+	mmcu?msp1=mmcu?msp430x4371 \
+	mmcu?msp2=mmcu?msp430x447 \
+	mmcu?msp2=mmcu?msp430x448 \
+	mmcu?msp2=mmcu?msp430x449 \
+	mmcu?msp3=mmcu?msp430xG4616 \
+	mmcu?msp3=mmcu?msp430xG4617 \
+	mmcu?msp3=mmcu?msp430xG4618 \
+	mmcu?msp3=mmcu?msp430xG4619 \
+	mmcu?msp4=mmcu?msp430x4783 \
+	mmcu?msp4=mmcu?msp430x4784 \
+	mmcu?msp4=mmcu?msp430x4793 \
+	mmcu?msp4=mmcu?msp430x4794 \
+	mmcu?msp5=mmcu?msp430x47166 \
+	mmcu?msp5=mmcu?msp430x47176 \
+	mmcu?msp5=mmcu?msp430x47186 \
+	mmcu?msp5=mmcu?msp430x47196 \
+	mmcu?msp5=mmcu?msp430x47167 \
+	mmcu?msp5=mmcu?msp430x47177 \
+	mmcu?msp5=mmcu?msp430x47187 \
+	mmcu?msp5=mmcu?msp430x47197 \
+	mmcu?msp6=mmcu?msp430x5418 \
+	mmcu?msp6=mmcu?msp430x5419 \
+	mmcu?msp6=mmcu?msp430x5435 \
+	mmcu?msp6=mmcu?msp430x5436 \
+	mmcu?msp6=mmcu?msp430x5437 \
+	mmcu?msp6=mmcu?msp430x5438 \
+	mmcu?msp6=mmcu?cc430x5123 \
+	mmcu?msp6=mmcu?cc430x5125 \
+	mmcu?msp6=mmcu?cc430x6125 \
+	mmcu?msp6=mmcu?cc430x6135 \
+	mmcu?msp6=mmcu?cc430x6126 \
+	mmcu?msp6=mmcu?cc430x5137 \
+	mmcu?msp6=mmcu?cc430x6127 \
+	mmcu?msp6=mmcu?cc430x6137
 
 MULTILIB_EXCEPTIONS =
 
@@ -150,4 +233,4 @@ msp430-function.o: $(srcdir)/config/msp430/msp430-function.c $(CONFIG_H) $(SYSTE
 	$(CC) -c $(ALL_CFLAGS) $(ALL_CPPFLAGS) $(INCLUDES) $(srcdir)/config/msp430/msp430-function.c
 
 
-	
\ No newline at end of file
+	
